reindexer/reindex_worker/src/main/scala/uk/ac/wellcome/platform/reindex_worker/services/ReindexService.scala

```diff
@@ -28,47 +30,58 @@ class ReindexService @Inject()(dynamoDbClient: AmazonDynamoDB,
     dynamoConfig = dynamoConfig
   )
 
-  def runReindex(reindexJob: ReindexJob): Future[List[Unit]] =
-    Future {
-      info(s"ReindexService running $reindexJob")
-
-      val table = Table[ReindexRecord](dynamoConfig.table)
+  def runReindex(reindexJob: ReindexJob): Future[List[Unit]] = {
+    info(s"ReindexService running $reindexJob")
+    val table = Table[ReindexRecord](dynamoConfig.table)
 
-      val index = table.index(indexName = dynamoConfig.index)
+    val outdatedRecordsFuture: Future[List[ReindexRecord]] = for {
+      index: SecondaryIndex[ReindexRecord] <- Future.fromTry(Try {
+        table.index(indexName = dynamoConfig.index)
+      })
 
       // We start by querying DynamoDB for every record in the reindex shard
       // that has an out-of-date reindexVersion.  If a shard was especially
       // large, this might cause out-of-memory errors -- in practice, we're
       // hoping that the shards/individual records are small enough for this
       // not to be a problem.
-      val results: List[Either[DynamoReadError, ReindexRecord]] =
+      results: List[Either[DynamoReadError, ReindexRecord]] <- Future {
         Scanamo.exec(dynamoDbClient)(
           index.query(
             'reindexShard -> reindexJob.shardId and
               KeyIs('reindexVersion, LT, reindexJob.desiredVersion)
           )
         )
+      }
 
-      val outdatedRecords: List[ReindexRecord] = results.map {
-        case Left(err: DynamoReadError) => {
-          warn(s"Failed to read Dynamo records: $err")
-          throw GracefulFailureException(
-            new RuntimeException(s"Error in the DynamoDB query: $err")
-          )
+      outdatedRecords: List[ReindexRecord] = results.map(extractRecord)
+    } yield outdatedRecords
+
+    // Then we PUT all the records.  It might be more efficient to do a
+    // bulk update, but this will do for now.
+    outdatedRecordsFuture.flatMap { outdatedRecords: List[ReindexRecord] =>
```

Comment:
What happened to the for comprehension?

Reply:
When I tried this earlier, I couldn’t find a way to put this inside the for comprehension that didn't cause the compiler to think I was returning Future[Nothing] or Future[Unit], rather than Future[List[Unit]].

Reply:
Let's go through it tomorrow, I think we should be able to use a for comprehension here. For now we can merge.